src/test/codewhisperer/util/authUtil.test.ts

```diff
@@ -0,0 +1,180 @@
+
+    return provider
+}
+
+function getTestTokenProvider(...[profile]: ConstructorParameters<typeof SsoAccessTokenProvider>) {
+    const key = getSsoProfileKey(profile)
+    const cachedProvider = tokenProviders.get(key)
+    if (cachedProvider !== undefined) {
+        return cachedProvider
+    }
+
+    const provider = createTestTokenProvider()
+    tokenProviders.set(key, provider)
+
+    return provider
+}
+
+let auth: Auth
+let store: ProfileStore
+let builderIdConn: SsoConnection | undefined
+let ssoConn: SsoConnection
+let entSsoConn: SsoConnection
+let mockConnListNoBuilder: Connection[]
+let mockConnListNoEntSso: Connection[]
+const ssoProfile = createSsoProfile()
+const awsBuilderIdProfile = createAwsBuilderIdProfile({ scopes: codewhispererScopes })
+const enterpriseSsoProfile = createEntSsoProfile({ scopes: codewhispererScopes })
+
+async function createMockConnections(){
+    store = new ProfileStore(new FakeMemento())
+    auth = new Auth(store, getTestTokenProvider, new CredentialsProviderManager())
+    builderIdConn = await auth.createConnection(awsBuilderIdProfile)
+    ssoConn = await auth.createConnection(ssoProfile)
+    entSsoConn = await auth.createConnection(enterpriseSsoProfile)
+}
+
+describe('AuthUtil', async function () {
+   before(async function () {
+       await createMockConnections()
+   })
+
+    beforeEach(async function () {
+        sinon.stub(Commands, 'register') 
+        store = new ProfileStore(new FakeMemento())
+        auth = new Auth(store, getTestTokenProvider, new CredentialsProviderManager())
+        
+        sinon.replace(SsoClient, 'create', () => {
+            const s = stub(SsoClient)
+            s.logout.resolves()
+
+            return s
+        })
+              
+    })
+
+    afterEach(async function () {
+        sinon.restore()
+    })
+
+    //TODO
+    //factory function to create connectionlists dynamically
+    //stub secondaryAuth useNewConnection? vs auth instance use connection
+    //remove auth and profile from beforeEach and see if it breaks
+    //todo: figure out what sinon.replace in beforeEach does
+    //add start url to params of create SSO profile
+    //figure out where to declare all the vars
+    //TODO add more intermittent asserts
+    it('if there is no valid AwsBuilderID conn, it will create one and use it', async function () {
+        mockConnListNoBuilder = [ssoConn]
+        sinon.stub(Auth.instance, 'listConnections').resolves(mockConnListNoBuilder)
+        sinon.stub(Auth.instance, 'createConnection').resolves(builderIdConn)
+        const authSpy = sinon.stub(Auth.instance, 'useConnection')
+
+        getTestWindow().onDidShowQuickPick(async picker => {
+            await picker.untilReady()
+            picker.acceptItem(picker.items[1])
+        })
+
+        await AuthUtil.instance.connectToAwsBuilderId()
+    
+        assert.ok(authSpy.called)
+        assert.strictEqual(authSpy.args[0][0], builderIdConn)
```

Comment:
Use `auth.activeConnection` instead of a spy

Reply:
Hi @JadenSimon , thanks for your comment!

I updated the tests to use `new Auth(auth)` per your note above and in doing so was able to remove the majority of the Auth stubs. 

However, I have still not found a way to implement the tests without the `useConnection` stub, due to the issue described below. Please let me know if the updated implementation works. 



**Detail:**
- When I use `AuthUtil(auth)` it creates the builderID connection, but generates the following error when the [secondaryAuth calls useConnection at the end of the function](https://github.com/aws/aws-toolkit-vscode/blob/master/src/codewhisperer/util/authUtil.ts#L95):

     - `Error: Connection does not exist: https://view.awsapps.com/start?scopes=codecatalyst:read_write,codewhisperer:analysis,codewhisperer:completions`

- It appears the `AuthUtil` is using the auth instance created in/for the test, but the `secondaryAuth` is not, and so it can’t find the profile associated with the new connection.

With that in mind:
- In the latest update, I used `new AuthUtil(auth)` as you mentioned and so was able to remove the stubs for `listConnection` and `createConnection`.  
- I still stub `useConnection`, in order to not get the above error and but still verify that the `AuthUtil` is calling it with the correct connection.

Does this work in your view? Alternatively, is there a good way to get the `secondaryAuth` to use the correct `Auth` obj (and by extension ProfileStore)?


Side note: 
- I am using ID as the connection property to check in the new assert, because when I try to use `type` or `label` I get the following error:
`Property 'type' does not exist on type 'Pick<IamConnection, "id”>’.` 
- I am not sure why `useConnection` is using `Pick<IamConnection, 'id'>` instead of `Pick<SsoConnection, 'id'>`

Reply:
> It appears the AuthUtil is using the auth instance created in/for the test, but the secondaryAuth is not, and so it can’t find the profile associated with the new connection.

Yup you're right. I've fixed this in https://github.com/aws/aws-toolkit-vscode/pull/3220/files#diff-b04b37a57656fe70020045378831b4b8df1ef84d02b18417f7b1599b2862cf1cR43-R48

> Does this work in your view?

So the thing with stubbing is that it _does_ work. There's nothing inherently wrong with using stubs and mocks. But each stubbed function or mocked object is a divergence from the original implementation. I think for our purposes stubbing `useConnection` is probably ok if my fix doesn't work.

> I am using ID as the connection property to check in the new assert
Are you looking at the captured param? The type signature is written intentionally loose but you can always cast it with `as Connection`. Also `Pick<IamConnection, 'id'>` is equivalent to `Pick<SsoConnection, 'id'>` in this context. I wouldn't worry too much about the exact names of things in TypeScript as it's a structurally-typed language. The only thing that matters are the shapes, which in this case is `{ id: string }` for both.